CIP-onchain-tx-chaining/README.md

```diff
@@ -0,0 +1,198 @@
+
+## Motivation: why is this CIP necessary?
+<!-- A clear explanation that introduces the reason for a proposal, its use cases and stakeholders. If the CIP changes an established design then it must outline design issues that motivate a rework. For complex proposals, authors must write a Cardano Problem Statement (CPS) as defined in CIP-9999 and link to it as the `Motivation`. -->
+
+The changes proposed in this CIP are implemented would allow for composability between transactions while preserving the deterministic behaviour.
+
+The main stakeholders would be dApp developers; who would have access to the transacionm chaining functionality also on-chain in plutus scripts, on top of the off-chain guarantee of the transactions validity.
+
+## Specification
+<!-- The technical specification should describe the proposed improvement in sufficient technical detail. In particular, it should provide enough information that an implementation can be performed solely on the basis of the design in the CIP. This is necessary to facilitate multiple, interoperable implementations. -->
+
+The core idea of the CIP is to add a new field in the transaction body representing the hash of the chained transaction (the transaction that uses the current transaction outputs as inputs).
+
+Despite being a simple idea it introduces a circular dependency of hashes;
+
+Infact the current transaction hash is used to describe the input used in the chained transaction;
+but the input of the chained transaction will determine the hash of the chained transaction;
+and the hash of the chained transaction would be included in the previous transaction body;
+and this would modify the hash of the initial transaction.
+
+(`current_tx_id` -> `utxo_input` -> `next_tx_id` -> `current_tx_id` )
+
+To break the circular dependecy we can modify the definition of a valid transaction input so that it doesn't require a transaciton hash if it is an output of a previus transaction.
+
+The only information needed for a chained input is the which previous transaction generated it and the output index.
+
+In particular the current `transacion_input` definition is
+
+```cddl
+transaction_input = [ transaction_id : $hash32
+                    , index : uint
+                    ]
+```
+
+We can define a new type of transaction input that here we call `chained_tx_input`
+
+```cddl
+chained_tx_input = [ prev_tx_pointer: uint
+                   , index : uint
+                   ]
+```
+
+so that we can introduce a new `chainable_transaction_input` cddl definition
+
+```cddl
+chainable_transaction_input = chained_tx_input / transaction_input
+```
+
+We can distinguish the two types of inputs based on the type of the first element of the array;
+
+if it is an unsigned integer it should be interpreted as a pointer to a previous transaction (`chained_tx_input` first element)
+(`0` being the transaction immediately before the current chained one)
```

Comment:
This is too vague: you are suggesting we index backwards from a transaction in _what_? The mempool? The chain? Both of those are created non-deterministically, so that seems extremely unlikely to work.

Reply:
I'm not really sure _where_ this should happen if you have any suggestions that would be great.

regarding both being not deterministic the pointer to the previous transaction only makes sense if there is a previous transaction pointing to the current one.

If none is present the transaction should be considered invalid (or at least wait to be processed until some other transaction includes the current as chained).

In this sense transactions are referencing chained ones forwards; UTxO is referencing backward.

Reply:
Then `prev_tx_pointer` is not needed. Output index is sufficient if we rely on the fact there is another transaction referening this one.